suro-core/src/main/java/com/netflix/suro/queue/MemoryQueue4Sink.java

```diff
@@ -20,14 +27,29 @@
     public static final String TYPE = "memory";
 
     private final BlockingQueue<Message> queue;
+    private final int capacity;
 
     @JsonCreator
     public MemoryQueue4Sink(@JsonProperty("capacity") int capacity) {
-        this.queue = new ArrayBlockingQueue(capacity);
+        this.capacity = capacity;
+        if (capacity == 0) {
+            this.queue = new SynchronousQueue();
+        } else {
+            this.queue = new ArrayBlockingQueue(capacity);
+        }
     }
 
     @Override
     public boolean offer(Message msg) {
+        if (capacity == 0) {
+            try {
+                // for synchronous queue, this will block until the sink takes the message from the queue
+                queue.put(msg);
+            } catch (InterruptedException e) {
+                return false;
+            }
+            return true;
+        }
         return queue.offer(msg);
```

Comment:
I thought you were just changing this offer to put. but that will be a big change.

you current change doesn't break any existing behavior. looks good/safe to me.

nitpicking: a if-else might be more straightforward.